### PR TITLE
 crimson/osd: add --mkkey support

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

File: src/common/buffer.cc [hunk 1]
```diff
@@ -1838,6 +1844,33 @@ ssize_t buffer::list::read_fd(int fd, size_t len)
   return ret;
 }
 
+#ifdef WITH_SEASTAR
```
Comment (tchaikov) [inline src/common/buffer.cc]: i can move this helper into `src/crimson/common` so we don't need to taint `buffer.h` and `buffer.cc` in this way.

File: src/ceph_osd.cc [hunk 2]
```diff
@@ -321,25 +321,21 @@ int main(int argc, const char **argv)
 
   if (mkkey) {
     common_init_finish(g_ceph_context);
-    KeyRing *keyring = KeyRing::create_empty();
-    if (!keyring) {
-      derr << "Unable to get a Ceph keyring." << dendl;
-      forker.exit(1);
-    }
+    KeyRing keyring;
```
Comment (rzarzynski) [inline src/ceph_osd.cc]: ACK.

File: src/crimson/osd/main.cc [hunk 3]
```diff
@@ -142,6 +167,9 @@ int main(int argc, char* argv[])
                                            hb_front_msgr.stop(),
                                            hb_back_msgr.stop());
         });
+        if (config.count("mkkey")) {
```
Comment (tchaikov) [inline src/crimson/osd/main.cc]: crimson-osd should right away if `--mkkey | --mkfs`. but in this change, we always expect `--mkfs` or none.

File: src/common/buffer.cc (hunk 1 above)
Comment (rzarzynski) [reply src/common/buffer.cc]: We might also consider `src/common/buffer_seastar.cc`.

File: src/crimson/osd/main.cc (hunk 3 above)
Comment (rzarzynski) [reply src/crimson/osd/main.cc]: `TODO` comment?

Comment (tchaikov) [reply src/crimson/osd/main.cc]: i will fix it in the next revision of this PR.

**Comment by tchaikov** [comment]
i will move the read/write file helpers to `src/common/buffer_seastar.cc` as `write(const bufferlist&, sstring path, permissions)` and `bufferlist read(sstring path)` are just consumers of `buffer::list`. there is no need to add them as member variables.

File: src/crimson/common/buffer_io.cc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <seastar/core/reactor.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/do_with.hh>
+
+#include "include/buffer.h"
+
+namespace ceph::buffer {
+
+seastar::future<> write_file(const ceph::buffer::list&& bl,
```
Comment (rzarzynski) [inline src/crimson/common/buffer_io.cc]: What's the reason for `const &&`?

Comment (tchaikov) [reply src/crimson/common/buffer_io.cc]: i failed to remove `const` when changing the type of `bl` from `const ceph::buffer::list&` to `ceph::buffer::list&&`, when i realized that it'd be easier to pass-by-move the buffer to the continuation of underlying `seastar::file`.

**Comment by tchaikov** [comment]
@rzarzynski  fixed and repushed.

**Pull request merged**